docs/intro/fastapi_gelai_searchbot.rst

```diff
@@ -0,0 +1,1556 @@
+        """
+        Recursively extract comments from a thread up to max_depth.
+        Returns a list of formatted comment strings.
+        """
+        if not comment or current_depth > max_depth:
+            return []
+
+        results = []
+
+        if comment["text"]:
+            timestamp = datetime.fromisoformat(comment["created_at"].replace("Z", "+00:00"))
+            author = comment["author"]
+            text = html.unescape(comment["text"])
+            formatted_comment = f"[{timestamp.strftime('%Y-%m-%d %H:%M')}] {author}: {text}"
+            results.append(("  " * current_depth) + formatted_comment)
+
+        if comment.get("children"):
+            for child in comment["children"][:max_children]:
+                child_comments = extract_comment_thread(child, max_depth, current_depth + 1)
+                results.extend(child_comments)
+
+        return results
+
+
+    def fetch_web_sources(query: str, limit: int = 5) -> list[WebSource]:
+        search_url = "http://hn.algolia.com/api/v1/search_by_date"
+
+        response = requests.get(
+            search_url,
+            params={
+                "query": query,
+                "tags": "story",
+                "hitsPerPage": limit,
+                "page": 0,
+            },
+        )
+
+        response.raise_for_status()
+        search_result = response.json()
+
+        web_sources = []
+        for hit in search_result.get("hits", []):
+            item_url = f"https://hn.algolia.com/api/v1/items/{hit['story_id']}"
+            response = requests.get(item_url)
+            response.raise_for_status()
+            item_result = response.json()
+
+            site_url = f"https://news.ycombinator.com/item?id={hit['story_id']}"
+            title = hit["title"]
+            comments = extract_comment_thread(item_result)
+            text = "\n".join(comments) if len(comments) > 0 else None
+            web_sources.append(
+                WebSource(url=site_url, title=title, text=text)
+            )
+
+        return web_sources
+
+
+    if __name__ == "__main__":
+        web_sources = fetch_web_sources("edgedb", limit=5)
+
+        for source in web_sources:
+            print(source.url)
+            print(source.title)
+            print(source.text)
+
+
+Notice that we've created another Pydantic type called ``WebSource`` to store
+our web search results. There's no framework-related reason for that, it's just
+nicer than passing dictionaries around.
+
+One more note: this snippet comes with an extra dependency called ``requests``,
+which is a library for making HTTP requests. Let's add it by running:
+
+.. code-block:: bash
+
+    $ uv add requests
+
+
+Now we can test our web search on its own by running it like this:
+
+
+.. code-block:: bash
+
+    $ python3 app/web.py
+
+
+It's time to reflect the new capabilities in our web app. We're going to hack
+our way around the fact that many stories on HN come with no comment threads.
```

Comment:
Maybe let's explain what we're doing so that no one is scared by "hacking around" and expecting something tricky